examples/utils/winit_app.rs

```diff
@@ -1,122 +1,151 @@
+        Self {
+            init,
+            init_surface,
+            _marker: PhantomData,
         }
     }
 
-    impl<T, Init, Handler> WinitApp<T, Init, Handler>
-    where Init: FnMut(&ActiveEventLoop) -> T,
-          Handler: FnMut(&mut T, Event<()>, &ActiveEventLoop)
+    /// Build a new application.
+    pub(crate) fn with_event_handler<F>(self, handler: F) -> WinitApp<T, S, Init, InitSurface, F>
+    where
+        F: FnMut(&mut T, Option<&mut S>, Event<()>, &ActiveEventLoop),
     {
-        /// Create a new application.
-        pub(crate) fn new(init: Init, event: Handler) -> Self {
-            Self {
-                init,
-                event,
-                state: None
-            }
-        }
+        WinitApp::new(self.init, self.init_surface, handler)
     }
+}
 
-    impl<T, Init, Handler> ApplicationHandler for WinitApp<T, Init, Handler>
-    where Init: FnMut(&ActiveEventLoop) -> T,
-          Handler: FnMut(&mut T, Event<()>, &ActiveEventLoop) {
-
-        fn resumed(&mut self, el: &ActiveEventLoop) {
-            debug_assert!(self.state.is_none());
-            self.state = Some((self.init)(el));
+impl<T, S, Init, InitSurface, Handler> WinitApp<T, S, Init, InitSurface, Handler>
+where
+    Init: FnMut(&ActiveEventLoop) -> T,
+    InitSurface: FnMut(&ActiveEventLoop, &mut T) -> S,
+    Handler: FnMut(&mut T, Option<&mut S>, Event<()>, &ActiveEventLoop),
+{
+    /// Create a new application.
+    pub(crate) fn new(init: Init, init_surface: InitSurface, event: Handler) -> Self {
+        Self {
+            init,
+            init_surface,
+            event,
+            state: None,
+            surface_state: None,
         }
+    }
+}
 
-        fn suspended(&mut self, _event_loop: &ActiveEventLoop) {
-            let state = self.state.take();
-            debug_assert!(state.is_some());
-            drop(state);
-        }
+impl<T, S, Init, InitSurface, Handler> ApplicationHandler
+    for WinitApp<T, S, Init, InitSurface, Handler>
+where
+    Init: FnMut(&ActiveEventLoop) -> T,
+    InitSurface: FnMut(&ActiveEventLoop, &mut T) -> S,
+    Handler: FnMut(&mut T, Option<&mut S>, Event<()>, &ActiveEventLoop),
+{
+    fn resumed(&mut self, el: &ActiveEventLoop) {
+        debug_assert!(self.state.is_none());
+        let mut state = (self.init)(el);
+        self.surface_state = Some((self.init_surface)(el, &mut state));
+        self.state = Some(state);
+    }
 
-        fn window_event(
-            &mut self,
-            event_loop: &ActiveEventLoop,
-            window_id: WindowId,
-            event: WindowEvent,
-        ) {
-            let state = self.state.as_mut().unwrap();
-            (self.event)(state, Event::WindowEvent {
-                window_id,
-                event
-            }, event_loop);
-        }
+    fn suspended(&mut self, _event_loop: &ActiveEventLoop) {
+        // TODO: Should we run a destruction function for the user?
```

Comment:
We are already destroying the surface, is there more to do here or what do you mean with "for the user"?

Reply:
We implicitly run `Drop` on it (and the `softbuffer::Surface` that we hold implements proper semantics), I don't think there's anything for the user to handle in here.

It's example code anyway, if users want to build off of this and need that functionality they can always add it themselves.